Improve waypoint movement

Stoabrogga opened the pull request:
## 🍰 Pullrequest
Adjust the algorithm for waypoint movement similar to CMaNGOS:
- Try to generate movement splines which are at least 6s long
- 1.5s before the spline is finished start a new spline

This prevents the animation resets occuring at each waypoint. I tested this with all kinds of different waypoint types with and without creature groups (see below). Please take a look at the code changes, I'm not a C++ professional.

### Issues
Closes #2713

### How2Test
Watch different creatures which use waypoint movement, e.g.:
- Thora Feathermoon (ID 5798, GUID 15169)
- Theresa (ID 5697, GUID 41840)
- Lord Kazzak (ID 12397, GUID 156950)
- Felguard Elite (ID 8717, GUID 2791)
- Pridewing Soarer (ID 6141, GUID 32374)
- Westfall Woodworker (ID 6670, GUID 45521)
- Thomas Booker (ID 1437, GUID 9468)
- Gordo (ID 10666, GUID 28413)
- Gazelle (ID 4166, GUID 14049)
- Wooly Kodo (ID 3237, GUID 15144)
- Warpwood Tangler (ID 11464, GUID 300943)
- Grunt Gargal (ID 9086, GUID 6883)

Also watch creature formations without waypoint movement, e.g. Emeraldon Boughguard (GUID 33246, ID 12474).

Stoabrogga committed:
Improve waypoint movement

Gamemechanicwow commented on src/game/Movement/WaypointMovementGenerator.cpp:
```diff
@@ -181,64 +203,111 @@ bool WaypointMovementGenerator<Creature>::OnArrived(Creature& creature)
-    if (WaypointPath const* pSubpath = (nextNode.path_id ? sWaypointMgr.GetPathFromOrigin(nextNode.path_id, 0, 0, PATH_FROM_SPECIAL) : nullptr))
+    PointsArray genPath; // will contain the generated path
+    genPath.reserve(20); // little optimization
+    genPath.push_back(Vector3(creature.GetPositionX(), creature.GetPositionY(), creature.GetPositionZ()));
```
Consider using emplace_back when applicable. Here it avoids an unnecessary copy.
```suggestion
    genPath.emplace_back(creature.GetPositionX(), creature.GetPositionY(), creature.GetPositionZ());
```

Gamemechanicwow commented on src/game/Movement/WaypointMovementGenerator.cpp:
```diff
@@ -137,16 +158,17 @@ bool WaypointMovementGenerator<Creature>::OnArrived(Creature& creature)
-    MANGOS_ASSERT(currPoint != i_path->end());
+    if (currPoint == i_path->end())
+        return false;
+    else
```
No need for else after a return

Gamemechanicwow commented on src/game/Movement/WaypointMovementGenerator.cpp:
```diff
@@ -181,64 +203,111 @@ bool WaypointMovementGenerator<Creature>::OnArrived(Creature& creature)
+                    ++loops;
+                }
+            }
+        }
```
If you avoid else after breaks, you can make the code a bit cleaner to look at.
```suggestion
        if (nextDelay)
            break;

        ++currPoint;

        if (currPoint == i_path->end()) 
        {
            if (!m_repeating)
                break;

            currPoint = i_path->begin();
            ++loops;
        }
```

Stoabrogga committed:
Remove unused variables

Stoabrogga committed:
Adjust direction calculation

ghost commented:
To be honest, I don't know if this is even worth to be merged. I installed it on my server and it looks better, but it makes the code more complicated and needs more calculations, so costs more performance, just for a small visual improvement.

mserajnik commented:
> To be honest, I don't know if this is even worth to be merged. I installed it on my server and it looks better, but it makes the code more complicated and needs more calculations, so costs more performance, just for a small visual improvement.

Since VMaNGOS tries to replicate Vanilla as closely as possible: if this is closer to the original blizz-like behavior than the current solution, it’s probably worth it, no?

Stoabrogga committed:
Bugfix spline point ID handling

ghost commented:
> Since VMaNGOS tries to replicate Vanilla as closely as possible: if this is closer to the original blizz-like behavior than the current solution, it’s probably worth it, no?

Yes, of course. There are probably also ways to improve performance, that's why I wrote that I'm not a C++ professional, just doing this as a hobby and wanted to share the stuff I played around with for the past few weeks. I also tested if there are any crashes by setting all creatures with WP movement active, preloading all maps and letting the server run for a few days.

Stoabrogga committed:
Feedback from Gamemechanic

ghost commented:
Thanks, Gamemechanic, for your feedback. I pushed a new commit with your suggestions.

Daribon commented:
Big improvement over current waypoint movement.

Here is before:
https://github.com/user-attachments/assets/14768b1f-64ba-4805-bf89-1c100940f9ce
And after:
https://github.com/user-attachments/assets/c4435f0c-2635-4409-9090-cd624c50437f

Do note that the NPC in above example is using custom waypoints, but it does a good job worst case scenario.
(I will add the sniffed waypoints when this is merged, since the animation would look ugly without this PR.)

Stoabrogga committed:
Merge branch 'vmdev' into wp_movement

Stoabrogga committed:
Fix patrol movement

Stoabrogga committed:
Fix special WP paths

Stoabrogga committed:
Fix patrol movement

If taking the points from the leader's WP movement ensure that the spline was created by WP movement generator (prevent a potential crash).

NickTyrer commented:
Can this be reopened? Seemed like a good PR
